File: client-desktop/src/services/callAudioService.ts
/**
 * Call Audio Service
 *
 * Handles playback of call-related audio sounds (ringtones, dial tones, etc.)
 */

export enum CallAudioType {
  DialTone = 'dial_tone',
  IncomingRingtone = 'ringtone_incoming',
  CallConnected = 'call_connected',
  CallEnded = 'call_ended',
  BusyTone = 'busy_tone',
}

const AUDIO_PATHS: Record<CallAudioType, string> = {
  [CallAudioType.DialTone]: '/audio/dial_tone.mp3',
  [CallAudioType.IncomingRingtone]: '/audio/ringtone_incoming.mp3',
  [CallAudioType.CallConnected]: '/audio/call_connected.mp3',
  [CallAudioType.CallEnded]: '/audio/call_ended.mp3',
  [CallAudioType.BusyTone]: '/audio/busy_tone.mp3',
};

const LOOPING_AUDIO = new Set([CallAudioType.DialTone, CallAudioType.IncomingRingtone]);

class CallAudioServiceClass {
  private dialToneAudio: HTMLAudioElement | null = null;
  private ringtoneAudio: HTMLAudioElement | null = null;
  private effectAudio: HTMLAudioElement | null = null;
  private currentPlaying: CallAudioType | null = null;

  constructor() {
    // Preload audio elements
    this.dialToneAudio = this.createAudio(CallAudioType.DialTone, true);
    this.ringtoneAudio = this.createAudio(CallAudioType.IncomingRingtone, true);
    this.effectAudio = this.createAudio(CallAudioType.CallConnected, false);
  }

  private createAudio(type: CallAudioType, loop: boolean): HTMLAudioElement {
    const audio = new Audio(AUDIO_PATHS[type]);
    audio.loop = loop;
    audio.preload = 'auto';
    return audio;
  }

  private getAudioForType(type: CallAudioType): HTMLAudioElement | null {
    switch (type) {
      case CallAudioType.DialTone:
        return this.dialToneAudio;
      case CallAudioType.IncomingRingtone:
        return this.ringtoneAudio;
      case CallAudioType.CallConnected:
      case CallAudioType.CallEnded:
      case CallAudioType.BusyTone:
        // Update effect audio source if needed
        if (this.effectAudio && this.effectAudio.src !== AUDIO_PATHS[type]) {
          this.effectAudio.src = AUDIO_PATHS[type];
        }
        return this.effectAudio;
      default:
        return null;
    }
  }

  /**
   * Play a specific audio type
   */
  public play(type: CallAudioType): void {
    // Stop any looping audio first
    if (LOOPING_AUDIO.has(type)) {
      this.stopLoopingAudio();
    }

    const audio = this.getAudioForType(type);
    if (!audio) return;

    // For effect audio, update the source
    if (!LOOPING_AUDIO.has(type)) {
      audio.src = AUDIO_PATHS[type];
    }

    audio.currentTime = 0;
    audio.play().catch((e) => {
      console.warn(`Failed to play audio ${type}:`, e);
    });

    if (LOOPING_AUDIO.has(type)) {
      this.currentPlaying = type;
    }
  }

  /**
   * Stop a specific audio type
   */
  public stop(type: CallAudioType): void {
    const audio = this.getAudioForType(type);
    if (audio) {
      audio.pause();
      audio.currentTime = 0;
    }

    if (this.currentPlaying === type) {
      this.currentPlaying = null;
    }
  }

  /**
   * Stop all looping audio (dial tone, ringtone)
   */
  private stopLoopingAudio(): void {
    if (this.dialToneAudio) {
      this.dialToneAudio.pause();
      this.dialToneAudio.currentTime = 0;
    }
    if (this.ringtoneAudio) {
      this.ringtoneAudio.pause();
      this.ringtoneAudio.currentTime = 0;
    }
    this.currentPlaying = null;
  }

  /**
   * Stop all audio playback
   */
  public stopAll(): void {
    this.stopLoopingAudio();
    if (this.effectAudio) {
      this.effectAudio.pause();
      this.effectAudio.currentTime = 0;
    }
  }

  /**
   * Set volume for all audio (0.0 to 1.0)
   */
  public setVolume(volume: number): void {
    const normalizedVolume = Math.max(0, Math.min(1, volume));
    if (this.dialToneAudio) this.dialToneAudio.volume = normalizedVolume;
    if (this.ringtoneAudio) this.ringtoneAudio.volume = normalizedVolume;
    if (this.effectAudio) this.effectAudio.volume = normalizedVolume;
  }

  // Convenience methods
  public playDialTone(): void {
    this.play(CallAudioType.DialTone);
  }

  public stopDialTone(): void {
    this.stop(CallAudioType.DialTone);
  }

  public playIncomingRingtone(): void {
    this.play(CallAudioType.IncomingRingtone);
  }

  public stopIncomingRingtone(): void {
    this.stop(CallAudioType.IncomingRingtone);
  }

  public playCallConnected(): void {
    this.play(CallAudioType.CallConnected);
  }

  public playCallEnded(): void {
    this.play(CallAudioType.CallEnded);
  }

  public playBusyTone(): void {
    this.play(CallAudioType.BusyTone);
  }
}

// Export singleton instance
export const CallAudioService = new CallAudioServiceClass();

File: client-mobile/lib/features/calls/data/services/call_audio_service.dart
/// Call Audio Service
///
/// Handles playback of call-related audio sounds (ringtones, dial tones, etc.)
library;

import 'dart:async';

import 'package:audioplayers/audioplayers.dart';
import 'package:logger/logger.dart';

/// Audio types for call sounds
enum CallAudioType {
  /// Dial tone when initiating an outgoing call
  dialTone,

  /// Incoming call ringtone
  incomingRingtone,

  /// Sound when call is connected
  callConnected,

  /// Sound when call ends
  callEnded,

  /// Busy tone when recipient is busy
  busyTone,
}

/// Service for managing call-related audio playback
class CallAudioService {
  final Logger _logger;
  final AudioPlayer _dialTonePlayer;
  final AudioPlayer _ringtonePlayer;
  final AudioPlayer _effectPlayer;

  bool _isInitialized = false;
  CallAudioType? _currentlyPlaying;

  /// Asset paths for audio files
  static const Map<CallAudioType, String> _assetPaths = {
    CallAudioType.dialTone: 'audio/dial_tone.mp3',
    CallAudioType.incomingRingtone: 'audio/ringtone_incoming.mp3',
    CallAudioType.callConnected: 'audio/call_connected.mp3',
    CallAudioType.callEnded: 'audio/call_ended.mp3',
    CallAudioType.busyTone: 'audio/busy_tone.mp3',
  };

  CallAudioService({
    required Logger logger,
    AudioPlayer? dialTonePlayer,
    AudioPlayer? ringtonePlayer,
    AudioPlayer? effectPlayer,
  }) : _logger = logger,
       _dialTonePlayer = dialTonePlayer ?? AudioPlayer(),
       _ringtonePlayer = ringtonePlayer ?? AudioPlayer(),
       _effectPlayer = effectPlayer ?? AudioPlayer();

  /// Initialize audio players
  Future<void> initialize() async {
    if (_isInitialized) return;

    try {
      // Set up dial tone player for looping
      await _dialTonePlayer.setReleaseMode(ReleaseMode.loop);

      // Set up ringtone player for looping
      await _ringtonePlayer.setReleaseMode(ReleaseMode.loop);

      // Set up effect player for one-shot playback
      await _effectPlayer.setReleaseMode(ReleaseMode.release);

      _isInitialized = true;
      _logger.i('CallAudioService initialized');
    } catch (e, stackTrace) {
      _logger.e(
        'Failed to initialize CallAudioService',
        error: e,
        stackTrace: stackTrace,
      );
    }
  }

  /// Play a specific audio type
  Future<void> play(CallAudioType type) async {
    if (!_isInitialized) {
      await initialize();
    }

    // Stop any currently playing sound of the same category
    await _stopCurrentIfSameCategory(type);

    try {
      final assetPath = _assetPaths[type];
      if (assetPath == null) {
        _logger.w('No asset path for audio type: $type');
        return;
      }

      final player = _getPlayerForType(type);
      await player.setSource(AssetSource(assetPath));
      await player.resume();

      _currentlyPlaying = type;
      _logger.d('Playing call audio: $type');
    } catch (e, stackTrace) {
      _logger.e(
        'Failed to play audio: $type',
        error: e,
        stackTrace: stackTrace,
      );
    }
  }

  /// Stop a specific audio type
  Future<void> stop(CallAudioType type) async {
    try {
      final player = _getPlayerForType(type);
      await player.stop();

      if (_currentlyPlaying == type) {
        _currentlyPlaying = null;
      }
      _logger.d('Stopped call audio: $type');
    } catch (e, stackTrace) {
      _logger.e(
        'Failed to stop audio: $type',
        error: e,
        stackTrace: stackTrace,
      );
    }
  }

  /// Stop all audio playback
  Future<void> stopAll() async {
    try {
      await Future.wait([
        _dialTonePlayer.stop(),
        _ringtonePlayer.stop(),
        _effectPlayer.stop(),
      ]);
      _currentlyPlaying = null;
      _logger.d('Stopped all call audio');
    } catch (e, stackTrace) {
      _logger.e('Failed to stop all audio', error: e, stackTrace: stackTrace);
    }
  }

  /// Set volume for all players (0.0 to 1.0)
  Future<void> setVolume(double volume) async {
    try {
      await Future.wait([
        _dialTonePlayer.setVolume(volume),
        _ringtonePlayer.setVolume(volume),
        _effectPlayer.setVolume(volume),
      ]);
    } catch (e) {
      _logger.e('Failed to set volume', error: e);
    }
  }

  /// Play dial tone for outgoing call
  Future<void> playDialTone() => play(CallAudioType.dialTone);

  /// Stop dial tone
  Future<void> stopDialTone() => stop(CallAudioType.dialTone);

  /// Play incoming call ringtone
  Future<void> playIncomingRingtone() => play(CallAudioType.incomingRingtone);

  /// Stop incoming ringtone
  Future<void> stopIncomingRingtone() => stop(CallAudioType.incomingRingtone);

  /// Play call connected sound
  Future<void> playCallConnected() => play(CallAudioType.callConnected);

  /// Play call ended sound
  Future<void> playCallEnded() => play(CallAudioType.callEnded);

  /// Play busy tone
  Future<void> playBusyTone() => play(CallAudioType.busyTone);

  /// Get the appropriate player for an audio type
  AudioPlayer _getPlayerForType(CallAudioType type) {
    switch (type) {
      case CallAudioType.dialTone:
        return _dialTonePlayer;
      case CallAudioType.incomingRingtone:
        return _ringtonePlayer;
      case CallAudioType.callConnected:
      case CallAudioType.callEnded:
      case CallAudioType.busyTone:
        return _effectPlayer;
    }
  }

  /// Stop current playback if same category (looping sounds)
  Future<void> _stopCurrentIfSameCategory(CallAudioType newType) async {
    if (_currentlyPlaying == null) return;

    // Stop dial tone when starting ringtone or vice versa
    if (newType == CallAudioType.dialTone &&
        _currentlyPlaying == CallAudioType.incomingRingtone) {
      await stop(CallAudioType.incomingRingtone);
    } else if (newType == CallAudioType.incomingRingtone &&
        _currentlyPlaying == CallAudioType.dialTone) {
      await stop(CallAudioType.dialTone);
    }
  }

  /// Dispose of all audio players
  Future<void> dispose() async {
    try {
      await Future.wait([
        _dialTonePlayer.dispose(),
        _ringtonePlayer.dispose(),
        _effectPlayer.dispose(),
      ]);
      _isInitialized = false;
      _logger.i('CallAudioService disposed');
    } catch (e, stackTrace) {
      _logger.e(
        'Failed to dispose CallAudioService',
        error: e,
        stackTrace: stackTrace,
      );
    }
  }
}
